src/Kinto.elm

```diff
@@ -0,0 +1,316 @@
+                joinUrl [ baseUrl, "buckets", bucketName, "collections", collectionName, "records", recordId ]
+
+
+client : Config -> Endpoint -> Verb -> Task Error Json.Value
+client config endpoint verb =
+    let
+        request =
+            { verb = verb
+            , url = endpointUrl config endpoint
+            , headers = config.headers
+            , body = Http.empty
+            }
+    in
+        (Http.send Http.defaultSettings request)
+            |> toKintoResponse
+
+
+withHeader : String -> String -> Config -> Config
+withHeader name value ({ headers } as config) =
+    let
+        newHeaders =
+            ( name, value ) :: headers
+    in
+        { config | headers = newHeaders }
+
+
+withAuthHeader : Auth -> Config -> Result Error Config
+withAuthHeader auth config =
+    case auth of
+        NoAuth ->
+            Ok config
+
+        Basic username password ->
+            let
+                hash str =
+                    Base64.encode str |> Result.formatError ConfigError
+
+                addAuthHeader hash =
+                    Ok (withHeader "Authorization" ("Basic " ++ hash) config)
+            in
+                hash (username ++ ":" ++ password)
+                    `Result.andThen` addAuthHeader
+
+        Bearer token ->
+            Ok (config |> withHeader "Authorization" ("Bearer " ++ token))
+
+
+configure : Url -> Auth -> Result Error Config
+configure baseUrl auth =
+    Config baseUrl []
+        |> withAuthHeader auth
+        |> Debug.log "Kinto request"
+
+
+
+-- Dealing with answers from the Kinto server
+
+
+bodyDataDecoder : Json.Decoder Json.Value
+bodyDataDecoder =
+    ("data" := Json.value)
+
+
+extractData : String -> StatusCode -> StatusMsg -> Task Error Json.Value
+extractData body statusCode statusMsg =
+    let
+        responseResult =
+            Json.decodeString bodyDataDecoder body
+    in
+        case responseResult of
+            Ok data ->
+                Task.succeed data
+
+            Err _ ->
+                -- Is it an ErrorRecord json?
+                decodeError body statusCode statusMsg
+
+
+errorDecoder : Json.Decoder ErrorRecord
+errorDecoder =
+    Json.object4 ErrorRecord
+        ("errno" := Json.int)
+        ("message" := Json.string)
+        ("code" := Json.int)
+        ("error" := Json.string)
+
+
+decodeError : String -> StatusCode -> StatusMsg -> Task Error a
+decodeError body statusCode statusMsg =
+    case Json.decodeString errorDecoder body of
+        Ok errorRecord ->
+            Task.fail (KintoError statusCode statusMsg errorRecord)
+
+        Err msg ->
+            Task.fail (ServerError statusCode statusMsg body)
+
+
+toKintoResponse : Task Http.RawError Http.Response -> Task Error Json.Value
+toKintoResponse response =
+    (Task.mapError NetworkError response)
+        `Task.andThen` (handleResponse extractData)
+
+
+handleResponse :
+    (String -> StatusCode -> StatusMsg -> Task Error Json.Value)
```

Comment:
First, it seems to me that `handleResponse` shouldn't perform any tasks. It should be pure, right? (Forgive me if I misunderstood, but I jumped to the conclusion that Elm's `Task` is like Haskell's `IO`, in that it indicates something with a side effect.)

There's two things that `handleResponse` does. One is it checks to see if the response was a `Text` response, and fails otherwise. The other is it calls `handle` in the "successful" case. But Elm already has a pretty common idiom for chaining "and if that worked, do this", and it's called `andThen`. At my last job I worked with @pjrt and he liked to argue that when your higher-order function is basically just a wrapper around `andThen`, you should stop making it be a higher-order function and just expose the `andThen`. I don't think this trick always works, but sometimes it does and it makes the resulting code feel more flexible and atomic. So what I'm suggesting is that if you have an `extractText` which does the pattern matching on the response and on success produces the arguments your `handle` function wants, then you can just use `andThen` explicitly, perhaps even in the callers of `handleResponse`.